core/src/main/java/com/linecorp/armeria/server/auth/AuthorizationStatus.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.auth;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+
+/**
+ * Provides a status of the request authorization operation, optionally combined with {@link AuthSuccessHandler}
+ * and {@link AuthFailureHandler} to facilitate custom status handling.
+ */
+@UnstableApi
+public interface AuthorizationStatus {
```

Comment:
Does `AuthorizationResult` sound better?

Reply:
As you wish. Maybe you should validate the design approach first and then perform some low-level brushing up of the API. Otherwise, I may be doing bunch of refactorings for something that could still be a through-ways code.
Do you agree with the API change in general, introducing `AuthorizationStatus` instead of the binary status?

Reply:
Your new API makes sense, so I just left the comments for minor stuff. Yeah, I agree with the idea.

I thought it may be better introducing a new interface that supercedes `Authorizer`, but perhaps we can live with the default method for now.

Could you just add some Javadoc in `Authorizer` that a user is expected to implement the new method and leave the old method empty, with an example?

Reply:
At second thought, what do you think about declaring a new interface? e.g.

- Add `interface AuthorizationFunction { CompletionStage<AuthorizationStatue> authorize(...) ... }`
- Deprecate `Authorizer` and add `Authorizer.toAuthorizationFunction()`
- Add various `AuthServiceBuilder.add*()` methods that accept `AuthorizationFunction`.
  - Note we can't add `add(Iterable<? extends AuthorizationFunction>)` though.
- Deprecate the `AuthServiceBuilder.add*()` methods that accept `Authorizer`.
- Modify `AuthServiceBuilder` and `AuthService` so they use `AuthorizationFunction` internally.

Reply:
Another question: What do you think about having three statuses?

- Authorized
- Unauthorized
- Unauthorized (Unrecognized or unsupported auth scheme, e.g. no 'Authorization' header)

Would the third status be useful for supporting multiple authorization mechanisms? We could add an option that tells `AuthService` to ignore the `AuthorizationFunction`s that returned the third status to support multiple authentication schemes. For example:

```java
// This will reject a request when:
// - Any of the three functions return 'Unauthorized' (2nd status).
// - All of the three functions return 'Unauthorized (unsupported)' (3rd status).
// and accept a request when:
// - None of the three functions return 'Unauthorized' (2nd status).
AuthService
  .builder()
  .add(oauth1func)
  .add(oauth2func)
  .add(basicFunc)
  .ignoreThirdStatus() // We need a better name.
  .build();

// If `ignoreThirdStatus()` is not called, the third status will be treated same as the second status.
```

Reply:
> At second thought, what do you think about declaring a new interface? e.g.
>     * Add `interface AuthorizationFunction { CompletionStage<AuthorizationStatue> authorize(...) ... }`
>     * Deprecate `Authorizer` and add `Authorizer.toAuthorizationFunction()`
>     ...

Looks like a massive refactoring! I don't see the end of this PR. Do you see specific reason to deprecate `Authorizer` as part of this?
I do feel that current authorization interfaces are somewhat limiting but we should think of the missing capabilities and design new interfaces around it.
Could you please check my question at the bottom regarding fine-grained access control.

Reply:
> Another question: What do you think about having three statuses?
>     * Authorized
>     * Unauthorized
>     * Unauthorized (Unrecognized or unsupported auth scheme, e.g. no 'Authorization' header)

Since the introduction of the `AuthorizationStatus` it's certainly possible to move away from the binary status. New statuses should even be backward compatible with binary status, just like I did with the handlers. However, it would be very difficult to keep backward compatibility semantically as it will contradict the presence of the third status. I thought you were against making such changes currently.

On the other hand, yes, it would be useful to have a special status to indicate that the `Authorizer` wants to ignore given authorization request. But how common this use case would be? I would question the need of `Authorizer` chaining in general. I think it's at least confusing but it might be dangerous in some cases. I don't see the need to have a chain of the authorizers to handle the same authorization scheme.

In order to support multiple authorization schemes, we should define clear and unambiguous map of Authorizers associated with different authorization schemes:
`AuthServiceBuilder.add(type, authorizer)` - maps an authorizer to a type (Basic, OAuth1a, OAuth2, ...); this should allow mapping to custom authorization schemes unknown to us.
`AuthServiceBuilder.setTypeDetector(typeDetector)` - to customize predefined type detection mechanism or to support custom authorization types.

Another authorization interfaces evolution I could think of related to fine-grained access control, where the standard authorization logic could be mapped to a particular `Route`.
`AuthServiceBuilder.addAccessRules(Route route, Iterable<String> roles)`
Additionally, we should define common policy for the routes no associated rules:
`AuthServiceBuilder.permitIfNoRulesDefined(boolean permit)`

Having this on the `AuthService` has much more sense than defining it for each authorization scheme based on `Authorizer`. All the context mapping would take place inside the `AuthService` and it would work across multiple schemes. Wile the `Authorizer` implements and extended method like below:
```
CompletionStage<AuthorizationStatus> authorizeAndSupplyHandlers(ServiceRequestContext ctx,
        @Nullable T data, Iterable<String> requiredRoles, boolean permitIfNoRulesDefined)
```

As I pointed, in future, we could add support [@javax.annotation.security.RolesAllowed](https://github.com/javaee/javax.annotation/blob/master/src/main/java/javax/annotation/security/RolesAllowed.java) annotation directly on the method signature (like most of JEE and JAX-RS frameworks do) like the following:
```
    @Put("/hello/{name}")
    @RolesAllowed ({"read", "write"})
    public HttpResponse putHello(@Param("name") String name) {
```
Having `AuthServiceBuilder.addAccessRules` allows us or the user translating the roles annotations to an access control policy - we could load it from a file, for instance.